Content.Server/Cargo/Systems/CargoSystem.Bounty.cs

```diff
@@ -0,0 +1,309 @@
+    {
+        foreach (var entry in entries)
+        {
+            var count = 0;
+
+            // store entities that already satisfied an
+            // entry so we don't double-count them.
+            var temp = new HashSet<EntityUid>();
+            foreach (var entity in entities)
+            {
+                if (!entry.Whitelist.IsValid(entity))
+                    continue;
+                count++;
+                temp.Add(entity);
+
+                if (count >= entry.Amount)
+                    break;
+            }
+
+            if (count < entry.Amount)
+                return false;
+
+            foreach (var ent in temp)
+            {
+                entities.Remove(ent);
+            }
+        }
+
+        return true;
+    }
+
+    [PublicAPI]
+    public bool TryAddBounty(EntityUid uid, StationCargoBountyDatabaseComponent? component = null)
+    {
+        // todo: consider making the cargo bounties weighted.
+        var bounty = _random.Pick(_protoMan.EnumeratePrototypes<CargoBountyPrototype>().ToList());
+        return TryAddBounty(uid, bounty, component);
+    }
+
+    [PublicAPI]
+    public bool TryAddBounty(EntityUid uid, string bountyId, StationCargoBountyDatabaseComponent? component = null)
+    {
+        if (!_protoMan.TryIndex<CargoBountyPrototype>(bountyId, out var bounty))
+        {
+            return false;
+        }
+
+        return TryAddBounty(uid, bounty, component);
+    }
+
+    public bool TryAddBounty(EntityUid uid, CargoBountyPrototype bounty, StationCargoBountyDatabaseComponent? component = null)
+    {
+        if (!Resolve(uid, ref component))
+            return false;
+
+        if (component.Bounties.Count >= component.MaxBounties)
+            return false;
+
+        var endTime = _timing.CurTime + _random.Pick(component.BountyDurations) + TimeSpan.FromSeconds(_random.Next(-10, 10));
+        component.Bounties.Add(new CargoBountyData(component.TotalBounties, bounty, endTime));
+        _adminLogger.Add(LogType.Action, LogImpact.Low, $"Added bounty \"{bounty.ID}\" (id:{component.TotalBounties}) to station {ToPrettyString(uid)}");
+        component.TotalBounties++;
+        return true;
+    }
+
+    [PublicAPI]
+    public bool TryRemoveBounty(EntityUid uid, int dataId, StationCargoBountyDatabaseComponent? component = null)
+    {
+        if (!TryGetBountyFromId(uid, dataId, out var data, component))
+            return false;
+
+        return TryRemoveBounty(uid, data.Value, component);
+    }
+
+    public bool TryRemoveBounty(EntityUid uid, CargoBountyData data, StationCargoBountyDatabaseComponent? component = null)
+    {
+        if (!Resolve(uid, ref component))
+            return false;
+
+        foreach (var bounty in new List<CargoBountyData>(component.Bounties))
+        {
+            if (bounty.Id == data.Id)
+            {
+                component.Bounties.Remove(data);
+                return true;
+            }
+        }
```

Comment:
If you just iterate, get the index, then RemoveAt, it saves allocating an entire list on the heap.